controller/sharding/cache.go

```diff
@@ -161,3 +173,60 @@ func (d *ClusterSharding) GetClusterAccessor() clusterAccessor {
 		return clusters
 	}
 }
+
+func (d *ClusterSharding) GetAppAccessor() appAccessor {
+	return func() []*v1alpha1.Application {
+		apps := make([]*v1alpha1.Application, 0, len(d.Apps))
+		for _, a := range d.Apps {
+			apps = append(apps, a)
+		}
+		return apps
+	}
+}
+
+func (sharding *ClusterSharding) AddApp(a *v1alpha1.Application) {
+	sharding.lock.Lock()
+	defer sharding.lock.Unlock()
+
+	_, ok := sharding.Apps[a.Name]
+	sharding.Apps[a.Name] = a
+	if !ok {
+		sharding.updateDistribution()
+	} else {
+		log.Debugf("Skipping sharding distribution update. App already added")
+	}
+}
+
+func (sharding *ClusterSharding) DeleteApp(a *v1alpha1.Application) {
+	sharding.lock.Lock()
+	defer sharding.lock.Unlock()
+	if _, ok := sharding.Apps[a.Name]; ok {
+		delete(sharding.Apps, a.Name)
+		sharding.updateDistribution()
+	}
+}
+
+func (sharding *ClusterSharding) UpdateApp(a *v1alpha1.Application) {
+	sharding.lock.Lock()
+	defer sharding.lock.Unlock()
+
+	_, ok := sharding.Apps[a.Name]
+	sharding.Apps[a.Name] = a
+	if !ok {
+		sharding.updateDistribution()
+	} else {
+		log.Debugf("Skipping sharding distribution update. No relevant changes")
+	}
+}
+
+func (sharding *ClusterSharding) GetAppDistribution() map[string]int {
+	appDistribution := make(map[string]int, len(sharding.Clusters))
```

Comment:
I believe this method should acquire a read lock from ClusterSharding.

Reply:
See above.

Reply:
Will need to be careful to add the read lock here as if someone calls GetAppDstribution inside a distributionFunction, the dead lock I mentioned above might occur. Unless my understanding is incorrect on the current implementation.

Reply:
Added a read lock.